[NEW] l10n_br_pos

mileo opened the pull request:
TODO:

res.partner
- [x] cnpj/cpf;
- [x] ie/rg;
- [x] suframa;
- [x] number;
- [x] district;
- [x] im;
- [x] relation with l10n_br_city;
- [x] relation with res.contry;

res.company:
- [x] cnpj/cpf;
- [x] ie/rg;
- [x] suframa;
- [x] number;
- [x] district;
- [x] im;
- [x] relation with l10n_br_city; @lfdivino 
- [x] relation with res.contry; @lfdivino 

Deseja CPF na nota? @lfdivino 
- [x] Add popup in cnpj/cpf in pos.order start;
- [x] OnChange res.partner if partner already registred;
- [x] pos.config: Sempre solicitar cnpj cpf ao iniciar uma nova venda;
- [x] Possible change cnpj/cpf of fiscal document in pos.order checkout;
  ![captura de tela de 2016-05-14 08 38 25](https://cloud.githubusercontent.com/assets/1975978/15268260/444bf772-19af-11e6-8a16-413776810b99.png)

Backport change cashier in pos.order:
- [x] Create a new module in oca/pos;
- [ ] Add it in l10n_br_pos dependencies ????? 
  ![captura de tela de 2016-05-14 08 37 36](https://cloud.githubusercontent.com/assets/1975978/15268259/3f0b3c64-19af-11e6-8487-6058653fa8eb.png)

Payment Term in Pos Order: @mileo
- [x] Create a new module in oca/pos;
- [x] Add it in l10n_br_pos dependencies ????? 
  ![selection](https://cloud.githubusercontent.com/assets/1975978/15268541/51bcccc4-19b9-11e6-9f1e-13c8e6cf11a4.png)

![captura de tela de 2016-05-14 08 36 27](https://cloud.githubusercontent.com/assets/1975978/15268254/ff89a1ac-19ae-11e6-86f5-62979af40932.png)

Others features:
- [x] Datetime in pos.order; CANCELED;

SAT/NFC-E @mileo  
- [ ] Test invoice and pos.order limit;
- [x] Save xml response of sat / nfc-e in pos.order;
- [x] When cancel a pos.oder create a nfc-e / sat cancel;
- [x] When return a pos.order create a nf-e;

ADD TESTS:
- [ ] create a partner with cpf in pos interface (P1);
- [ ] create a partner with cnpj in pos interface (P2);
- [ ] create a pos.order sat;
- [ ] create a pos.order sat for partner (P1) ;
- [ ] create a pos.order sat for partner (P2) ;
- [ ] create a pos.order nfc-e;
- [ ] create a pos.order nfc-e for partner (P1) ;
- [ ] create a pos.order nfc-e for partner (P2) ;
- [ ] create a pos.order nf-e;
- [ ] create a pos.order nf-e for partner (P1) ;
- [ ] create a pos.order nf-e for partner (P2) ;
- [ ] create a pos.order cancel for sat;
- [ ] create a pos.order cancel for nfc-e;
- [ ] create a pos.order cancel for nf-e;
- [ ] create a pos.order refund for sat;
- [ ] create a pos.order refund for nfc-e;
- [ ] create a pos.order refund for nf-e;

mstuttgart committed:
[FIX] Export tags adjust

mileo commented on l10n_br_pos/static/src/xml/pos.xml:
```diff
@@ -0,0 +1,137 @@
+<templates id="template" xml:space="preserve">
+    <t t-extend="ClientLine">
+        <t t-jquery="tr.client-line > td:first-child" t-operation="after">
+            <td><t t-esc='partner.cnpj_cpf or ""'/></td>
```
@lfdivino Vamos ter que seguir a mesma abordagem

 or ""' Visto que alguns campos estão aprecendo como False na tela e quando tentatamos salvar um cliente sem IE/RG temos um erro de validação que inclusive nem aparece.

mileo commented:
CC: @leorochael @danimaribeiro @gwmoura @renatonlima @rvalyi 
@carlosmaguda

mileo commented:
![captura de tela de 2016-05-16 06 02 28](https://cloud.githubusercontent.com/assets/1975978/15285131/da6536f6-1b2b-11e6-8dfd-991c19641c3f.png)

@leorochael @lfdivino It may be faster to develop in this way

mileo commented:
@gwmoura

moylop260 commented:
Hello, we have added lint checks for js
Could you review them to see if are fails valid for us?

``` bash
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:19:109: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:34:29: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:38:17: Bad line breaking before '&&'.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:39:17: Bad line breaking before '&&'.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:46:30: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:50:17: Bad line breaking before '&&'.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:51:17: Bad line breaking before '&&'.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:67:22: ['simplified'] is better written in dot notation.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:72:22: ['simplified'] is better written in dot notation.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:85:39: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:94:23: Use '!==' to compare with 'undefined'.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:106:52: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:116:44: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:117:30: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:133:55: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:135:44: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:138:44: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:139:30: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/main.js:146:9: Expected an assignment or function call and instead saw an expression.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/db.js:46:49: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/db.js:49:60: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/db.js:65:21: 're' used out of scope.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/db.js:75:7: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/models.js:74:47: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7903(javascript-lint), ] static/src/js/models.js:77:26: Missing semicolon.
l10n_br_pos/__init__.py:1: [W7908(missing-newline-extrafiles), ] views/pos_template.xml Missing newline
************* Module l10n_br_pos.models.pos_order
l10n_br_pos/models/pos_order.py:27: [W8104(api-one-deprecated), PosOrder.action_invoice] api.one deprecated
```

https://travis-ci.org/OCA/l10n-brazil/jobs/130205994#L889

mileo commented:
@moylop260 Thanks for your tip, I will check it today.

rvalyi commented:
terrific! I didn't review the code yet, but that's a very nice contrib!

rvalyi commented:
@lfdivino wanted to give your work a try on the OCA Runbot so I went here:
https://runbot.odoo-community.org/runbot/build/3149105
but there is a strange error: pos_pricelist module not found (the same if you try to connect and install l10n_br_pos manually. At first you might think about about the pos repo would be missing in oca_dependencies.txt, but it seems it's there. So at the end I'm not sure why it happens and when the ci/runibot check still seems green at the same time... Any idea?

lfdivino commented:
@rvalyi 

Already exist an issue about runbots not installing l10n_br module created by Danimar:

https://github.com/OCA/l10n-brazil/issues/296

I think this is not a specifc error.

mileo committed:
[NEW] Add l10n_br_pos

lfdivino committed:
Refatorada funcao para cancelar cupom fiscal

lfdivino committed:
Refatorada funcao refund() para realizar o cancelamento de uma pos order

lfdivino committed:
Adicionado pop-up para avisar que a venda foi cancelada corretamente e tambem adicionado configuracao de ambiente (Homologacao/Producao) no res_company e no pos_config

Conflicts:
	l10n_br_pos/static/src/js/devices.js

lfdivino committed:
[ADD] Configuracoes do ambiente SAT no res_company e um grupo de configuracoes do SAT no pos_config. Todas estas configuracoes ja esta sendo coletadas pelo javascript do POS.

lfdivino committed:
Pequenas correcoes em strings

mileo committed:
[REF] Refatoracao dos campos e strings

mileo committed:
[FIX] ambiente required

mileo committed:
[ADD] Printer parms

mileo committed:
[FIX] cnpj software house in json

lfdivino committed:
Adicionado funcao para dar o start no SAT e na impressora, adicionado campos relacionados ao cancelamento da pos_order e refatorado a funcao de cancelar a ultima order

Conflicts:
	l10n_br_pos/views/point_of_sale_view.xml

lfdivino committed:
[ADD] Adicionado funcao para reimprimir o retorno do Cfe da venda

mileo commented:
@lfdivino Acredito que demos um force com a origem errada.

lfdivino committed:
[FIX] correcoes nas funcoes de reimpressao e de cancelamento do Cfe

lfdivino committed:
[FIX] Refatorado funcoes de cancelamento da ultima venda e de reimpressao, adicionado coluna informando o status das ultimas vendas do pos

lfdivino committed:
Refatorado funcao para salvar e setar novos usuarios cadastrando o cpf ou cnpj

lfdivino committed:
[FIX] Retornar cpf_cnpj do cliente selecionado no json que vai para o lancamento do CFe

fernandomr committed:
modo de pagamento e credenciadora de cartao

mileo committed:
[NEW] Allowed CFOP

hendrixcosta committed:
[ADD] Search by chave_CFE / CNPJ/CPF

mileo committed:
[NEW] Add product origin and fiscal_classification_id

mileo committed:
[NEW] Add fiscal_category on pos.config

mileo committed:
[Add] New fields on frontend

mileo committed:
[REF][ADD] Payment terms/mode fields

mileo committed:
[FIX] Export for printing

mileo committed:
[REF] Refatoracao

mileo committed:
[NEW] Pos Order Refund

mileo committed:
[FIX] License

mileo committed:
[ADD] Pos default fiscal category

mileo committed:
[NEW] l10n_br_tax.estimate in l10n_br_pos

mileo committed:
[NEW] l10n_br_tax.estimate in l10n_br_pos

mileo committed:
[NEW] Copy product estimated taxes to pos

lfdivino committed:
Adicionado field para a assinatura do SAT no pos.config e arrumado funcao modelo para carregar os partners com o field cnpj_cpf

lfdivino committed:
Minor fixes

rvalyi commented:
@mileo I saw you merged the POS in your kmee fork, does it mean this PR is now ready for review instead of work in progress? Thanks.

lfdivino committed:
[ADD] cpf do destinatario do CFe para realizar o cancelamento do Cupom

lfdivino committed:
[FIX] Consertada funcao para cadastrar e setar novos clientes pelo cpf e arrumado a view do product_template com relacao ao field estimated_taxes

lfdivino committed:
[FIX] Alterado atributo store do field estimated_taxes

lfdivino committed:
[FIX] Ao cadastrar novo cpf nao altera mais a lista de preco, continua a que esta setada para o POS

lfdivino committed:
[FIX] Funcao que valida o troco no POS

lfdivino committed:
[FIX] Correcao na funcao que cria a pos_order de cancelamento, agora criando um pagamento negativando a pos_order original, e alterado o name da pos_order de cancelamento de REFUND para CANCEL

lfdivino committed:
[FIX] Adicionado campos chave_cfe e total na tabela de ultimas vendas e removido o campo nome

lfdivino committed:
Set up partner pressing enter after insert cpf/cnpj number in the text input, disabled buttons change price and set discount, connect sat after pos initiated

fernandomr committed:
FIX partner creation

sadamo committed:
[NEW] created new related document for SAT refund

lfdivino committed:
Adicionado funcao para exportar os xmls das vendas realizadas no POS em um determinado periodo

lfdivino committed:
[FIX] Modificado caminho em que e criado o zip com os xmls

lfdivino committed:
Setado pasta temporaria para a geracao do zip com as xmls do sat

lfdivino committed:
Adicionado verificacao para adicionar o name de uma determinada empresa caso a mesma nao seja uma empresa pai

lfdivino committed:
Adicionado verificacao para buscar somente as pos_orders que possuem um arquivo de retorno e tambem adicionado os xmls de cancelamento

lfdivino committed:
[FIX] codigo que gera o zip com os xmls do sat de uma determinada empresa

lfdivino committed:
[FIX] Funcao que busca qual o zip deve ser exibido para download do usuario

lfdivino committed:
[FIX] codigo para salvar o zip_sat_file no objeto 'nfe.xml.periodic.export'

sadamo committed:
Added the pos_order_picking_link in l10n_br_pos dependecies and altered the l10n_br_account_product dependence to l10n_br_sale_stock

sadamo committed:
Extends the funcionality of the referenced document in refund invoice to pos orders (sat-cfe)

lfdivino committed:
[FIX] Extensão das telas para adicionar corretamente a tela da Lista das ultimas vendas e refatorada a função de cadastro de novos usuarios para eliminar os caracteres não numéricos.

sadamo committed:
Invoice related documents onchange methods to clear wrong infos and some small fixes

mileo committed:
Replace tabs with spaces

find . -name '*.xml' ! -type d -exec bash -c 'expand -t 4 "$0" > /tmp/e && mv /tmp/e "$0"' {} \;

mileo committed:
OCA clean headers

for i in `find ./ -iname '*.py'`; do sed -i ':a;N;$!ba;s|#                                                                             #\n# This program is free.*#\n###############################################################################|# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html|g' $i; done

for i in `find ./ -iname '*.py'`; do sed -i ':a;N;$!ba;s|#\n# This program.*###############################################################################|# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html|g' $i; done

for i in `find ./ -iname '*.py'`; do sed -i ':a;N;$!ba;s|###############################################################################\n#                                                                             #||g' $i; done

for i in `find ./ -iname '*.py'`; do sed -i ':a;N;$!ba;s|###############################################################################\n#||g' $i; done

lfdivino committed:
Refactory funções validate_order(), export_as_JSON(), export_for_print()

lfdivino committed:
Adicionado regra para verificar se existe cadastrada uma empresa pai, em caso positivo seta o id desta empresa pai no cadastro do usuario para generaliza-lo

lfdivino committed:
[REF] Refatorado funcao para criar o botao 'ultimas vendas'

mileo committed:
[FIX] Removing computed field from l10n_br_pos

sadamo committed:
[fix] Acesso do campo fiscal_document_access_key e corte do 'CFe' no inicio da chave para preencher o campo documentos relacionados

sadamo committed:
Copy the partner_id of pos_order to the wizard of devolution

lfdivino committed:
correcao no texto de impressao da view

lfdivino committed:
[FIX] Funcoes que geram o json ao realizar o pagamento da order. Corrigido funcao que verifica se o cpf/cnpj ja esta cadastrado em um partner, o numero do documento com e sem pontuacao.

lfdivino committed:
[FIX] montagem das orderlines no json para enviar ao posbox do SAT

lfdivino committed:
[FIX] Estimated_taxes e remocao da pontuacao do cpf_cnpj para enviar para o sat

lfdivino committed:
[FIX] Verificar se a venda possui cliente ou nao

lfdivino committed:
[FIX] Faltava o atributo 'origin' ao mondar as lines de produtos do pedido

mileo commented:
Rebased!

rvalyi commented:
@mileo is it ready for review or still under heavy WIP?

mileo commented:
1. Adiciona as funcionalidades desejadas para o POS no Brasil;
2. Permite a seleção entre opções de envio SAT e NFC-E;
3. Permite a seleção entre opções de impressora:
   - Epson TM-T20;
   - Bematech MP4200TH;
   - Daruma DR700;
   - Elgin I9;

Testado e em Produção
- Simples Nacional , SP
- Bematech MP4200TH
- SAT: Tanca e Bematech 1000;

mileo commented:
@rvalyi Já esta em produção desta forma, funcionando com SAT, portanto ser revisado sim.

Tem vários PR pra OCA que ajudam ele a funcionar, seria legal revisar eles tb.

https://github.com/OCA/pos/pull/122
https://github.com/OCA/pos/pull/114
https://github.com/OCA/pos/pull/106

@lfdivino esta faltando o do vendedor né?

Estou verificando com o time a possibilidade de fazermos um Hackathon nos próximos dias para colocar o NFC-E.

E ainda falta o hw_sat que não esta nesse PR, que contem a o serviço e a lib de transmissão do SAT.

mileo committed:
[FIX] add stock_picking_invoice_link to l10n_br_pos dependencies

gglimass commented:
Este projeto fora abandonado? já existe alguma versão para a disponibilização para a comunidade?

mileo commented:
Não não foi abandonado. Estamos usando em produção desde agosto.

Precisamos de revisores apenas

gglimass commented:
Quando e onde esta feature será disponibilizada para a comunidade, já que se trata de programa livre?

danimaribeiro commented:
13 de maio de 2016, neste mesmo local :)

leorochael commented:
@gglimass, é bom lembrar que, "programa livre" não obriga ninguém a entregar nada para a comunidade. 

A licença GPL e derivadas obrigam apenas o código fonte do programa a ser compartilhado junto com os binários (o que, em um código Python, normalmente já acontece automaticamente), e restringe o distribuidor de processar o recipiente caso o recipiente redistribua o código recebido para outros.

Software Livre não é um privilégio dos usuários, é um presente de quem doa. Quer ver ele sendo doado mais rápido? Então ajude a comunidade contribuindo e, mais importante, testando e fazendo revisões do código que foi contribuído.

gglimass commented:
Você esta enganado amigo... se o software é colocado como livre ele tem q cumprir premissas básicas que são:
Todo e qualquer programa de computador cuja licença de direito de autor conceda ao utilizador as seguintes quatro liberdades é considerado software livre:

Liberdade 0: A liberdade de executar o programa, para qualquer propósito.
Liberdade 1: A liberdade de estudar como o programa funciona, e adaptá-lo para as suas necessidades. Acesso ao código-fonte é um pré-requisito para esta liberdade.
Liberdade 2: A liberdade de redistribuir cópias de modo que você possa ajudar ao seu próximo.
Liberdade 3: A liberdade de aperfeiçoar o programa, e redistribuir seus aperfeiçoamentos, de modo que toda a comunidade possa se beneficiar.
SE NÃO É DIVULGADO NÃO É PROGRAMA LIVRE.. ALIAS O  O ODOO BRASIL É UM SOFTWARE LIVRE DE FAZ DE CONTA, SEU CODIGO É DITO LIVRE POREM APENAS UM PEQUENO GRUPO PODE USA-LO. UMA VERGONHA

gglimass commented:
só completando..... programa livre nunca foi um presente, existe muitas vantagens de se desenvolver programa livre.

zimbora69 commented:
@mileo 

Boa tarde,

Primeiramente parabéns pela iniciativa neste projeto, quero ajudar no projeto.

Sabe-me dizer se o SAT Bematech RB-2000 (têm wi-fi) funciona com a vossa biblioteca. 

[Sat Cfe]
(https://github.com/kmee/satcfe)

mileo commented:
@zimbora69 Acredito que sim, pois o RB-1000 funciona e a dll é do mesmo fabricante.

Depois que vc ligou aqui na empresa tentei te mandar uma msg no telegram. Acredito que vc não recebeu.

zimbora69 commented:
@mileo 

Bom dia,

Tou com um erro na instalação do l10n_br_pos.

Erro: http://pastebin.com/tBPcsmGi

mileo commented:
Sua versão da localização tem este trecho de código? 

https://github.com/OCA/l10n-brazil/search?utf8=%E2%9C%93&q=l10n_br_account_product_nfe_form

mileo commented:
Qual quer coisa vc pode tentar com o nosso fork

https://github.com/kmee/l10n-brazil/tree/8.0

Pode ser que o PR não esteja com as ultimas modificações. 

@lfdivino  sabe me dizer se o PR esta atualizado?

zimbora69 commented:
@mileo 

Já consegui, tou agora a realizar testes.

Obrigado

mileo commented:
@zimbora69 preciso subir o modulo hw_sat que deve rodar no hardware. 

Faço isso hoje ainda

mileo commented:
@zimbora69  https://github.com/kmee/odoo-brazil-posbox

mileo commented:
@renatonlima Da para integrar o repo https://github.com/kmee/odoo-brazil-posbox aqui ou deixamos em uma projeto separado?

zimbora69 commented:
@mileo 

Em termos de hardware é preciso o SAT, impressora e já era? Ou preciso comprar o PosBox ou o Raspberry Pi

mileo commented:
Vc pode rodar ele no linux, sobe um odoo com aquele buildout e instala do SAT como de costume no linux, define o caminho da DLL no pos.config e a impressora dentre as suportadas.

Se vc for rodar eme num rasperberry pi, vc vai ter que analisar se existe driver para o ARM, que pelo que eu sei são apenas alguns fabricantes que já disponibilizaram isso.

No seu caso ref a Bematech, vc vai precisar usar linux ou windows. ( O posbox para windows ainda estamos trabalhando nele para ficar suave a instalação @ananiasfilho pode te dar mais detalhes)

E qual a impressora vc esta usando?

mileo commented:
Explicando melhor: o caixa deverá ser linux e estar rodando o buildout do posbox.

zimbora69 commented:
Impressora: Bematech MP4200TH

mileo commented:
Qual é o tipo da conexão dela? USB, Serial ?

zimbora69 commented:
Conexão é USB

zimbora69 commented:
@mileo 
O SAT path é o caminho do .dll do SAT?

E o Printer parameters serve para que?

mileo commented:
![pos config](https://cloud.githubusercontent.com/assets/1975978/24117178/f4ff3cce-0d87-11e7-8047-2608931c12ab.png)

Segue a configuração
